CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

IntegralProgrammer opened the pull request:
This PR adds a project-specific _CARDS is down_ page that is returned to the client by the reverse proxy whenever the CARDS Docker container is unavailable.

To test:

- Build this branch with `mvn clean install`.
- `cd compose-cluster`
- For each project in `cards4care`, `cards4kids`, `cards4lfs`, `cards4proms`:
  - `python3 generate_compose_yaml.py --dev_docker_image --oak_filesystem --cards_project <PROJECT NAME> && docker-compose build && docker-compose up -d`, replacing `<PROJECT NAME>` with the actual project name (eg. `cards4care`).
  - Wait for CARDS to become available at `http://localhost:8080`
  - `docker-compose stop cardsinitial`
  - Visiting `http://localhost:8080` should now show the project-specific _down for maintenance_ page.
  - `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
- Also, attempt the above with a generic CARDS deployment (ie. don't specify the `--cards_project` parameter)
- Verify that this works in all 4 proxy modes: HTTP, HTTP+SAML, HTTPS, HTTPS+SAML.

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

If the proxy container cannot connect to the cardsinitial container, serve the file proxyerror/503.html instead of showing the default Apache HTTP 503 Service Unavailable error page

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Display the project-specific logo on the custom HTTP 503 error page

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Make the error message displayed specific to the CARDS project - eg. "WilliamsDB is down for maintenance" or "DATA-PRO is down for maintenance"

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Add some styling to the 503 error page

TODO: Vertically center the content on the page

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Vertically center the content on the page

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Fix the file paths to logo_light_bg.png in compose-cluster/generate_compose_yaml.py

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Change width of project-specific logo

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Fix file path for generic CARDS logo in compose-cluster/generate_compose_yaml.py

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Fix compose-cluster/proxy/http_saml_000-default.conf

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Fix compose-cluster/proxy/https_saml_000-default.conf

sdumitriu commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -111,6 +111,33 @@ def generateSelfSignedCert():
 
+def getCardsProjectLogoPath(project_name):
+  logoPathMap = {}
+  logoPathMap['cards4care'] = "cardiac-rehab-resources/clinical-data/src/main/media/SLING-INF/content/libs/cards/resources/media/care/logo_light_bg.png"
+  logoPathMap['cards4kids'] = "kids-resources/clinical-data/src/main/media/SLING-INF/content/libs/cards/resources/media/kids/logo_light_bg.png"
+  logoPathMap['cards4lfs'] = "lfs-resources/clinical-data/src/main/media/SLING-INF/content/libs/cards/resources/media/lfs/logo_light_bg.png"
+  logoPathMap['cards4proms'] = "proms-resources/clinical-data/src/main/media/SLING-INF/content/libs/cards/resources/media/proms/logo_light_bg.png"
```
I really don't like this, but there's no easy way to do it better. A reorganization of the codebase is needed...

marta- replied:
Is there a way this code could somehow leverage the config jsons we have for each project ([example](https://github.com/data-team-uhn/cards/tree/dev/proms-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/conf)) and read the app name and logo path from there?

marta- replied:
Something like:
* `project_id` = strip the "cards4" prefix from `project_name`
* load `{{project_id}}-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/conf/AppName.json` parse it and set `project_name` = the `AppName` field (defaulting to CARDS if the json is not found or the AppName field is not found in it)
* load `{{project_id}}-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/conf/Media.json`, extract the `logoLight` field and set `projectLogoPath` to be `{{project_id}}-resources/clinical-data/src/main/media/SLING-INF/content/{{logoLight}}`, defaulting to the CARDS logo if anything goes wrong along the way (the json is not found, the field is not found in it, or the path leads to a file that does not exist).

It's not super elegant and relies quite a bit on our convention for organizing files in child projects, but at least we wouldn't have to modify this script when adding a new project (such as `heracles`).

Thoughts? @IntegralProgrammer @sdumitriu

marta- replied:
Note: the above approach won't work for `cards4care`/`cardiac-rehab-resources` and will end up defaulting to the CARDS logo and app name, which is OK since that project is not maintained on the main branch anyway.

marta- committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down
Minor styling adjustments for the placeholder page

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Modify getCardsProjectLogoPath so that it attempts to obtain the project-specific logo from a PROJECT_NAME-resources directory

This way, generate_compose_yaml.py does not need to be modified whenever a new project is added

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Modify getCardsApplicationName so that it attempts to obtain the project application name from a PROJECT_NAME-resources directory

This way, generate_compose_yaml.py does not need to be modified whenever a new project is added

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Refactor by introducing the getPathToProjectResourcesDirectory method

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Improve the getPathToProjectResourcesDirectory to handle the mapping of cards4care --> ../cardiac-rehab-resources/

marta- commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -111,6 +111,105 @@ def generateSelfSignedCert():
+  if project_name in resourcesPathMap:
+    return resourcesPathMap[project_name]
+
+  if not project_name.startswith(CARDS4_PREFIX):
+    return None
```
In `start_cards.sh` we can specify the project without `cards4`, e.g. `start_cards.sh -P lfs`. Can we do the same here?

marta- replied:
This is actually out of scope, since short form project names are not supported at all at this time by docker deployment.

IntegralProgrammer committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Handle the case where --cards_project is not given

IntegralProgrammer commented:
Improved code in `generate_compose_yaml.py` to address https://github.com/data-team-uhn/cards/pull/1049#discussion_r887293073. Tested all 4 CARDS projects (Cards4Care, Cards4Kids, Cards4LFS, Cards4Proms) as well as the generic CARDS. Screenshots are below.

![cards4care_is_down](https://user-images.githubusercontent.com/5349728/171935571-57a62440-a6ae-4b6b-abdb-712048476bde.png)

![cards4kids_is_down](https://user-images.githubusercontent.com/5349728/171935840-8cae3b6e-cc08-4dca-b2f0-ba38781ba0eb.png)


![cards4lfs_is_down](https://user-images.githubusercontent.com/5349728/171935638-46768054-c950-4c33-a56d-827396318d61.png)
375ee7b6f8.png)

![cards4proms_is_down](https://user-images.githubusercontent.com/5349728/171935675-b02bf0af-907f-463c-8af5-18d1bf53e4f2.png)

![cards_is_down](https://user-images.githubusercontent.com/5349728/171935692-d2bd365c-202e-400b-a7a7-c57e2fe209f7.png)

marta- reviewed (approved):
The code looks good to me. Didn't test with docker.

marta- committed:
CARDS-1726: Create a simple placeholder html page template to be displayed when CARDS is down

Switched  from `font-family: sans` to `font-family: sans-serif` so that Windows browsers can understand it too.

marta- commented:
> > Tested with docker, looks good to me. One minor thing: I seem to be getting a different default font
> 
> Hm, I don't like this font, we need to make sure it's `sans`.

Fixed in 0181b34.

marta- commented:
Tested, all looks good.
